src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,413 @@
+        /// But if this creation is correct, then all subsequent uses are correct.
+        /// </summary>
+        /// <param name="ptr">An unmanaged pointer to memory.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="T"/> is reference type or contains pointers and hence cannot be stored in unmanaged memory.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        public unsafe Span(void* pointer, int length)
+        {
+            if (!SpanHelpers.IsReferenceFree<T>())
+                throw new ArgumentException(SR.Format(SR.Argument_InvalidTypeWithPointersNotSupported, typeof(T)));
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Length = length;
+            _object = null;
+            _byteOffset = new IntPtr(pointer);
+        }
+
+        /// <summary>
+        /// Create a new span over a portion of a regular managed object. This can be useful
+        /// if part of a managed object represents a "fixed array." This is dangerous because 
+        /// "length" is not checked, nor is the fact that "rawPointer" actually lies within the object.
+        /// </summary>
+        /// <param name="obj">The managed object that contains the data to span over.</param>
+        /// <param name="rawPointer">A reference to data within that object.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when the specified object is an array or string. Use the type-specific constructor
+        /// for these special cases.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when the specified object is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        public static Span<T> DangerousCreate(object obj, ref T rawPointer, int length)
+        {
+            if (obj == null)
+                throw new ArgumentNullException(nameof(obj));
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            // Disallow "special" objects that have their own object headers and would yield unexpected (and runtime-dependent) results.
+            // We already have constructors for arrays (and that enforce the passing of the properly typed single-dimensional array.)
+            // This constructor is not useful for strings (cannot express"ref s[i]").
+            if (obj is Array)
```

Comment:
string: an immutable type being wrapped inside a mutable Span... and you can't take a "ref s[0]" so there's no correct way to use it anyway.

array: this boils down to - do we want to allow multidim arrays or not? (what guarantees do we make about the layout of those (row order or column order?) We could allow T[] to go through but that's another check and it duplicates functionality that we have more strongly typed constructors for. 

The offered up rationale for these were plain old objects with "fixed array-like stuff in them" so that's what I'm scoping these to.